### PR TITLE
[EC-502] Rate Limiting Improvements

### PR DESCRIPTION
**Pull request by shane-melton**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
There were some concerns with the original implementation of distributed IP rate limiting in #2060 ([PS-93](https://bitwarden.atlassian.net/browse/PS-93)).

It also helps improve performance by using in-memory stores for the Client and Ip Policy stores as they are not currently in use and would add extra calls to Redis. ([EC-503](https://bitwarden.atlassian.net/browse/EC-503))

Additionally it adds some configuration options to `GlobalSettings.cs` for configuring/tweaking the distributed rate limiting processing strategy. It has some initial default values, but further discussion would be beneficial to be sure they're most effective.

## Code changes

* **src/Core/Utilities/CustomRedisProcessingStrategy.cs:** The processing strategy that will gracefully handle a disconnected or degraded Redis connection.

* **src/SharedWeb/Utilities/ServiceCollectionExtensions.cs:** 
  * Check for new `DistributedIpRateLimiting.Enabled` flag. 
  * Manually register the policy stores to use in-memory stores as we don't need them in the distributed cache. 
  * Register the `CustomRedisProcessingStrategy`

* **src/Core/Settings/GlobalSettings.cs:** Add a new `DistributedIpRateLimitingSettings` class/property for configuring the distributed IP rate limiting and graceful fallback. 

* **appsettings.json:** Add default values for the new `DistributedIpRateLimitingSettings`

* **test/Core.Test/Utilities/CustomRedisProcessingStrategyTests.cs:** Unit tests for the new processing strategy.
## Before you submit

```
- [X] I have checked for formatting errors (`dotnet format --verify-no-changes`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [X] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: src/Core/Utilities/CustomRedisProcessingStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                    "Redis timeout threshold has been exceeded, backing off and skipping IP rate limiting");
+                return SkipRateLimitResult();
+            }
+        }
+
+        var counterId = BuildCounterKey(requestIdentity, rule, counterKeyBuilder, rateLimitOptions);
+
+        try
+        {
+            return await IncrementAsync(counterId, rule.PeriodTimespan ?? rule.Period.ToTimeSpan(), _config.RateIncrementer);
```
Comment (kspearrin) [inline src/Core/Utilities/CustomRedisProcessingStrategy.cs:75]: can we just call `base.ProcessRequestAsync` here instead?

File: src/Core/Utilities/CustomRedisProcessingStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+            _memoryCache.Set(_redisTimeoutCacheKey, timeoutCounter,
+                new MemoryCacheEntryOptions { AbsoluteExpiration = timeoutCounter.ExpiresAt });
+
+            // Just because Redis timed out does not mean we should kill the request
+            return SkipRateLimitResult();
+        }
+    }
+
+    private async Task<RateLimitCounter> IncrementAsync(string counterId, TimeSpan interval,
```
Comment (kspearrin) [inline src/Core/Utilities/CustomRedisProcessingStrategy.cs:95]: Do you need this if we just call `base` like suggested above?

File: src/Core/Utilities/CustomRedisProcessingStrategy.cs (hunk 1 above)
Comment (shane-melton) [reply src/Core/Utilities/CustomRedisProcessingStrategy.cs]: Yep, that does make a lot more sense

File: src/Core/Utilities/CustomRedisProcessingStrategy.cs (hunk 2 above)
Comment (shane-melton) [reply src/Core/Utilities/CustomRedisProcessingStrategy.cs]: Nope, we no longer need it if using `base`

**Review by kspearrin — APPROVED** [review]

**Review by mimartin12 — APPROVED** [review]

**Pull request merged**